frontend/src/components/tab/tab.tsx

```diff
@@ -0,0 +1,90 @@
+/* eslint-disable react-refresh/only-export-components */
+import styled from '@emotion/styled'
+import {createContext, PropsWithChildren, useContext, useState} from 'react'
+
+interface TabContext {
+  value: number
+  setValue: React.Dispatch<React.SetStateAction<number>>
+}
+
+const tabContext = createContext<TabContext | null>(null)
+
+interface ProviderProps {
+  defaultValue: number
+}
+
+function TabProvider({defaultValue, children}: PropsWithChildren<ProviderProps>) {
+  const [value, setValue] = useState(defaultValue)
+  return <tabContext.Provider value={{value, setValue}}>{children}</tabContext.Provider>
+}
+
+type TabOnChange = (value: number) => void
+const onChangeContext = createContext<TabOnChange | null>(null)
+
+interface TabItemListProps {
+  onChange?: TabOnChange
+}
+
+const ItemListContainer = styled.div({display: 'flex'})
+
+function TabItemList({children, onChange}: PropsWithChildren<TabItemListProps>) {
+  return (
+    <onChangeContext.Provider value={onChange ?? null}>
+      <ItemListContainer>{children}</ItemListContainer>
+    </onChangeContext.Provider>
+  )
+}
+
+interface ItemContainerProps {
+  selected?: boolean
+}
+
+const ItemContainer = styled.div<ItemContainerProps>(({theme, selected}) => ({
+  borderBottomWidth: '0.2rem',
+  borderColor: selected ? theme.color.primary : 'transparent',
+  borderStyle: 'solid',
+  textAlign: 'center',
+  height: '48px',
+  lineHeight: '46px',
+  color: theme.color.gray700,
+  flexGrow: 1,
+  fontSize: '1.4rem',
+}))
+
+interface ItemProps {
+  label: string
+  value: number
+}
+
+function TabItem({label, value}: ItemProps) {
+  const context = useContext(tabContext)
+  const onChange = useContext(onChangeContext)
+  const isSelected = value === context?.value
+  return (
+    <ItemContainer
+      selected={isSelected}
+      onClick={() => {
+        if (!isSelected) onChange?.(value)
+        context?.setValue(value)
```

Comment:
```suggestion
        context?.setValue(prev => {
          if(prev !== value) onChange?.(value)
          return value
        })
```

이렇게도 작성해볼 수 있을거 같아요!

Reply:
저도 원래 이렇게 작성했었는데 isSelected 변수를 새로 만들면서 코드를 수정했어요
위와 같이 코드를 작성해야 하는 이유는 setter가 비동기로 실행되기 때문인가요?

Reply:
그런 의도로 말씀드린건 아니었고 위와 같이도 작성할 수 있어서 보여드렸어요!